{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " \n",
    "# <center> Stats on possibilities with Ingrahams dataset </center>\n",
    "\n",
    "**Takes:** J. Ingrahams datasplit file from (1) which is based on the structural database CATH4.2 (2) and use this for a basis for splitting  data for this simulation dataset.\n",
    "In addition, it extends the filtering of PDBs as most proteins in Ingraham's already prefiltered CATH4.2 dataset are not useable for simulations. The filtering of pdbs is based on the below filtering conditions. \n",
    "    \n",
    "**Outputs**  a json `CATH4.2_INITIAL_STATUS.json` that shows the state of all proteins in the Ingraham data split and has the below format.\n",
    "It is additionally updated in the `data/CATH4.2_ALL_STATUS.json` by the notebook `track_simulations.ipynb` with information of which proteins has successfully been simulated or not : \n",
    "    \n",
    "```\n",
    "# The dictionary in the json has the format: \n",
    "dictionary = {pdbID:  {\n",
    "                'cath_nodes': str, \n",
    "                'size': int, # len of residues \n",
    "                'method':str #'NMR' or 'X-ray',\n",
    "                'chains':int, \n",
    "                'split':str, # 'train', 'test' or 'validation'\n",
    "                'prefilter status': str # 'removed' or 'kept',\n",
    "                'prefilter reasons': str # 'N.A', 'resolution' 'membrane protein', ... etc,\n",
    "                'simulation status':str #'N.A', 'finished' or 'unsuccessfull',\n",
    "                'unsuccessfull reason':str #'>RMSD', 'NaN'...etc\n",
    "               }}\n",
    "\n",
    "```\n",
    "\n",
    "    \n",
    "The conditions used to filter out proteins are as follows: \n",
    "\n",
    "    - is a membrane protein\n",
    "    - unsuitable experimental method  for solving the structure (unsuitable only in the sense of usefull for simulations)\n",
    "    - low X-ray resolution \n",
    "    - to high RMSD (>3.5) between models in NMR models \n",
    "    - contains unknown amino acids\n",
    "    - gaps/unresolved parts inside the chain of a protein (in terminals these are allowed)\n",
    "    - unsuitable ligand(s) (unsuitable only in the sense of usefull for automation of simulations)\n",
    "    - too big a protein (>500 residues)\n",
    "    - any chain too small in the  protein (<50 residues). Meaning it is unlikely to process a (somewhat) 3D structure by itself\n",
    "\n",
    "**OBS!** It should be stressed that neither of the filtering conditions descriped above necessarily say anything about the quality of the given structure, but are here used as a way to avoid issues when automating simulations. \n",
    "                                               \n",
    "\n",
    "```\n",
    "(1) \n",
    "@inproceedings{ingraham2019generative,\n",
    "author = {Ingraham, J. and Garg, V. K. and Barzilay, R. and Jaakkola, T.},\n",
    "title = {Generative Models for Graph-Based Protein Design},\n",
    "journal = {Advances in Neural Information Processing Systems}\n",
    "year = {2019}\n",
    "}\n",
    "``` \n",
    "\n",
    "``` \n",
    "(2) \n",
    "@article{CATH,\n",
    "author = {Sillitoe, I. and Bordin, N. and Dawson, N. and Waman, V. P. and Ashford, P. and Scholes, H. M. and Pang, C. S. M. and Woodridge, L. and Rauer, C. and Sen, N. and Abbasian, M. and Le Cornu, S. and Lam, S. D. and Berka, K. and V., Ivana H. and Svobodova, R. and Lees, J. and Orengo, C. A.},\n",
    "title = \"{CATH: increased structural coverage of functional space}\",\n",
    "journal = {Nucleic Acids Research},\n",
    "year = {2020},\n",
    "doi = {10.1093/nar/gkaa1079}}\n",
    "```\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys,os,json,time, gzip, time\n",
    "import numpy as np\n",
    "\n",
    "# Bio packages\n",
    "import mdtraj as md\n",
    "import Bio.PDB.mmtf as BioMMTF\n",
    "from Bio.PDB.Polypeptide import PPBuilder\n",
    "import mmtf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STATUS OF FILTERING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "OVERVIEW:\n",
      "=============\n",
      "\n",
      "Total pdbs in Ingraham's original dataset:\n",
      "\t 18783\n",
      "\n",
      "Total number of pdbs  which are filtered out from simulation dataset:\n",
      "\t 16588 \n",
      "\n",
      "Total number of pdbs  which are deemed initially suitable to keep in simulation dataset:\n",
      "\t 2195 \n",
      "\n",
      "\n",
      "\n",
      "REASONS for FILTERING OUT PDBS :\n",
      "================================\n",
      "\tObsolete in pdb database : 7\n",
      "\tMembrane protein: 286\n",
      "\tUnsuitable experimental method : 4\n",
      "\tResolution of X-ray structure to high: 2649\n",
      "\tRMSD between NMR models >3.5 : 1718\n",
      "\tContains unknown amino acid : 130\n",
      "\tAny one chain in structure >50 residues : 990\n",
      "\tGap in chain (unresolved parts in the middel of the structure) : 8435\n",
      "\tUsuitable ligand for automatic simulation : 7700\n",
      "\tTotal number of residues >500 : 10164\n",
      "\tTotal number of residues <50 : 95\n",
      "\tError in reading pdb: 23\n"
     ]
    }
   ],
   "source": [
    "with open('CATH4.2_INITIAL_STATUS.json', 'r') as fp:\n",
    "    init_data = json.load(fp)  \n",
    "\n",
    "print('\\n\\nOVERVIEW:\\n=============')    \n",
    "print(f\"\\nTotal pdbs in Ingraham's original dataset:\\n\\t {len(init_data)}\")\n",
    "\n",
    "filtered_out = [pdb for pdb in init_data if init_data[pdb]['prefilter status']=='removed']\n",
    "print(f\"\\nTotal number of pdbs  which are filtered out from simulation dataset:\\n\\t {len(filtered_out)} \")\n",
    "simulation_suitable = [pdb for pdb in init_data if init_data[pdb]['prefilter status']=='kept']\n",
    "print(f\"\\nTotal number of pdbs  which are deemed initially suitable to keep in simulation dataset:\\n\\t {len(simulation_suitable)} \")\n",
    "\n",
    "\n",
    "# prefilter reasons \n",
    "print('\\n\\n\\nREASONS for FILTERING OUT PDBS :\\n================================')    \n",
    "\n",
    "obsolete =  [pdb for pdb in init_data if 'obsolete' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tObsolete in pdb database : {len(obsolete)}\")\n",
    "\n",
    "membranes =  [pdb for pdb in init_data if 'membrane protein' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tMembrane protein: {len(membranes)}\")\n",
    "\n",
    "exp =  [pdb for pdb in init_data if 'unsuitable exp method' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tUnsuitable experimental method : {len(exp)}\")\n",
    "\n",
    "resolution =  [pdb for pdb in init_data if 'low resolution' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tResolution of X-ray structure to high: {len(resolution)}\")\n",
    "\n",
    "NMR =  [pdb for pdb in init_data if 'NMR rmsd to high' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tRMSD between NMR models >3.5 : {len(NMR)}\")\n",
    "\n",
    "aa =  [pdb for pdb in init_data if 'unknown_aa' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tContains unknown amino acid : {len(aa)}\")\n",
    "\n",
    "chain =  [pdb for pdb in init_data if 'one chain too short' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tAny one chain in structure >50 residues : {len(chain)}\")\n",
    "\n",
    "gap =  [pdb for pdb in init_data if 'gap in chain' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tGap in chain (unresolved parts in the middel of the structure) : {len(gap)}\")\n",
    "\n",
    "ligand =  [pdb for pdb in init_data if 'unsuitable ligand' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tUsuitable ligand for automatic simulation : {len(ligand)}\")\n",
    "\n",
    "big =  [pdb for pdb in init_data if 'too big' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tTotal number of residues >500 : {len(big)}\")\n",
    "\n",
    "small =  [pdb for pdb in init_data if 'too small' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tTotal number of residues <50 : {len(small)}\")\n",
    "\n",
    "error =  [pdb for pdb in init_data if 'error in reading pdb' in init_data[pdb]['prefilter reasons']]\n",
    "print(f\"\\tError in reading pdb: {len(error)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions used to filter out pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "def load_membrane_protein_list():\n",
    "    '''\n",
    "    Read membrane protein database into list of pdb IDs.\n",
    "    Using Stephen White's MPstruc database for finding membrane proteins: \n",
    "    https://blanco.biomol.uci.edu/mpstruc/?expandTblOnLoad=1\n",
    "    load xml over all protein structures that are membrane proteins\n",
    "    '''\n",
    "    \n",
    "    import xml.etree.ElementTree as ET\n",
    "    mpstruc = '/home/trz846/MD_ML/CATH4.3/helper_files/membrane_protein_database_mpstruc.xml'\n",
    "    tree = ET.parse(mpstruc)\n",
    "    root = tree.getroot()\n",
    "    # load pdb IDs of all  membrane proteins\n",
    "    membrane_proteins = [pdb.text for pdb in root.iter('pdbCode')]\n",
    "    \n",
    "    \n",
    "    return membrane_proteins\n",
    "\n",
    "\n",
    "\n",
    "def membrane_protein(pdb_name):\n",
    "    '''check if categorised as membrane protein in MPStruc'''\n",
    "    \n",
    "    pdb_name = pdb_name.upper() \n",
    "    if pdb_name in membrane_proteins:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def unsuitable_experimental_method(mmtf_record):\n",
    "    '''check that method used to solve protein structure is allowed''' \n",
    "    # define allowed methods == good for MD simulations\n",
    "    allowed_exp = ['ELECTRON MICROSCOPY', \n",
    "                    'SOLID-STATE NMR', \n",
    "                    'SOLUTION NMR',\n",
    "                    'X-RAY DIFFRACTION',\n",
    "                    'ELECTRON CRYSTALLOGRAPHY']\n",
    "\n",
    "    # get list of methods used for solving structure\n",
    "    used_exp = mmtf_record.experimental_methods\n",
    "    \n",
    "    # see if any of them is among suitable. \n",
    "    is_allowed = [exp for exp  in used_exp if exp in allowed_exp]\n",
    "\n",
    "    if  is_allowed: return False\n",
    "    else:  return True \n",
    "\n",
    "\n",
    "def low_resolution(mmtf_record):\n",
    "    '''check if resolution lower than 2.5 AA if experimental method is x-ray crystallization'''\n",
    "    \n",
    "    try: # can only get resolution for some exp methods \n",
    "        resolution = mmtf_record.resolution \n",
    "        if resolution < 2.5: return False \n",
    "        elif resolution >2.5: return True\n",
    "        else: return False\n",
    "    \n",
    "    except: # lazy as ... I know... \n",
    "        return False \n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "def polymeric(biopython_structure):\n",
    "    count_chains = len(biopython_structure)\n",
    "    return count_chains\n",
    "\n",
    "\n",
    "def aa_selenocystein(mmtf_record):\n",
    "    '''Test if selenocystein is in the protein'''\n",
    "    protein_entities = [entity for entity in  mmtf_record.entity_list if entity['type'] == 'polymer']\n",
    "    for chain in protein_entities:\n",
    "        sequence = chain['sequence']\n",
    "        seleno_cys_present = False\n",
    "        if 'U' in sequence:\n",
    "            seleno_cys_present = True\n",
    "            break\n",
    "    \n",
    "    return seleno_cys_present \n",
    "\n",
    "\n",
    "def unknown_aa(mmtf_record):\n",
    "    '''filters out proteins structures that contains amino acids that are not \n",
    "    uniqly determined, such as X,J,B etc\n",
    "    '''\n",
    "    \n",
    "    known_aa = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S'\n",
    "                ,'T','Y','V','W', 'U'] # U = Selenocystein.\n",
    "    \n",
    "    bad_aa = False\n",
    "    protein_entities = [entity for entity in  mmtf_record.entity_list if entity['type'] == 'polymer']\n",
    "    for chain in protein_entities:\n",
    "        sequence = chain['sequence']\n",
    "        bad_aa = [aa for aa in sequence if aa not in known_aa]\n",
    "        if len(bad_aa) > 0: \n",
    "            bad_aa = True\n",
    "            break \n",
    "            \n",
    "    if bad_aa: return True\n",
    "    else: return False\n",
    "    \n",
    "    \n",
    "    \n",
    "def missing_aa_inside_protein(biopython_structure):\n",
    "    '''check if any missing/unsolved residues in structure in the \n",
    "    middle of the sequence. Checks based on structure not on jumps\n",
    "    in sequence (the latter can happen when lab cuts out part of seq)\n",
    "    ''' \n",
    "    \n",
    "    ppb = PPBuilder()\n",
    "    for i, chain in enumerate(biopython_structure):\n",
    "        pps = ppb.build_peptides(chain)\n",
    "        if len(pps) > 1:\n",
    "            # chain breaks in structure (not in sequence) if missing residues in pps => pps>1\n",
    "            return True\n",
    "\n",
    "        elif len(pps)==1:\n",
    "            return False\n",
    "\n",
    "\n",
    "    \n",
    "def too_big(biopython_structure, max_len):\n",
    "    '''checks if protein is larger than defined maximum lenght (in residues)\n",
    "    '''\n",
    "    \n",
    "    size = get_size(biopython_structure)\n",
    "    if size > max_len: return True\n",
    "    else: return False\n",
    "    \n",
    "    \n",
    "def too_small(biopython_structure, min_len):\n",
    "    '''checks if protein is smaller than defined minimum lenght (in residues)\n",
    "    and thus will not have any structure in inself. \n",
    "    '''\n",
    "    \n",
    "    size = get_size(biopython_structure)\n",
    "    if size < min_len: return True\n",
    "    else: return False\n",
    "\n",
    "    \n",
    "def get_size(biopython_structure):\n",
    "    '''get total residue size of all chains in protein\n",
    "    ''' \n",
    "    size = 0\n",
    "    for chain in biopython_structure:\n",
    "        size+=len(chain)\n",
    "    return size\n",
    "\n",
    "\n",
    "def polymeric_with_single_chain_too_short(biopython_structure):\n",
    "    '''returns false if polymeric i.e. multiple chains \n",
    "    OBS! does not work for NMR - it reads each model as 1 chain \n",
    "    '''\n",
    "    \n",
    "    # if homo-polymer then each of the identical chains are put in 1 entity, i.e. 1 entity contains all homomers in pdb\n",
    "    any_chain_too_short = False\n",
    "    for chain in biopython_structure:\n",
    "        if len(chain) < 50:\n",
    "            any_chain_too_short = True\n",
    "    return any_chain_too_short\n",
    "\n",
    "\n",
    "\n",
    "def get_biopython_structure(mmtf_record):\n",
    "    '''Uses biopython loader to get (first) model in pdb structure\n",
    "    Needed for some processing.\n",
    "    '''\n",
    "    \n",
    "    structure = BioMMTF.get_from_decoded(mmtf_record)\n",
    "    first_model = structure.get_list()[0]\n",
    "    return first_model\n",
    "\n",
    "\n",
    "def unsuitable_ligands(mmtf_record):\n",
    "    '''Check which ligands exist in pdb structure and assess whether \n",
    "    structure should be discarded based on it\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    standard_crystallization_ligands = [\n",
    "        'water', 'ACETATE ION', 'ACETIC ACID',\"ADENOSINE-5'-DIPHOSPHATE\",\n",
    "        'ALPHA-D-MANNOSE','AMMONIUM ION','S-ADENOSYL-L-HOMOCYSTEINE',\n",
    "        \"ADENOSINE MONOPHOSPHATE\",\"S-ADENOSYLMETHIONINE\",\"BENZOIC ACID\",\n",
    "        'CHLORIDE ION', 'CITRIC ACID',\"CITRATE ANION\",\"BETA-CAROTENE\",\n",
    "        \"COENZYME A\",\"ACETYL COENZYME *A\",'DI(HYDROXYETHYL)ETHER' ,\n",
    "        \"ETHANOL\",'GLYCEROL', \"ALPHA-D-GLUCOSE\", \"HEXAETHYLENE GLYCOL\",\n",
    "        \"GLYCINE\",'BETA-D-GLUCOSE',\"2-(ACETYLAMINO)-2-DEOXY-A-D-GLUCOPYRANOSE\",\n",
    "        \"BETA-D-GALACTOSE\",\"GUANOSINE-5'-DIPHOSPHATE\",\"GUANOSINE-5'-TRIPHOSPHATE\",\n",
    "        \"GLUTAMIC ACID\",'FLAVIN-ADENINE DINUCLEOTIDE','FLAVIN MONONUCLEOTIDE',\n",
    "        'FORMIC ACID',\"HEME\",'ALPHA-L-FUCOSE','IMIDAZOLE',\"ISOPROPYL ALCOHOL\",\n",
    "        'BETA-D-MANNOSE','BETA-MERCAPTOETHANOL','MALONATE ION', \"MALONIC ACID\",\n",
    "        'D-MALATE','N-ACETYL-D-GLUCOSAMINE' ,'NICOTINAMIDE-ADENINE-DINUCLEOTIDE',\n",
    "        \"NADPH DIHYDRO-NICOTINAMIDE-ADENINE-DINUCLEOTIDE PHOSPHATE\",\"NITRATE\",\n",
    "        'PENTAETHYLENE GLYCOL','PHOSPHATE ION',\"PYRIDOXAL-5'-PHOSPHATE 161\",\n",
    "        \"OXYGEN MOLECULE\",'SULFATE ION',\"SUCROSE\",\"SUCCINIC ACID\",\n",
    "        'TETRAETHYLENE GLYCOL', 'TRIETHYLENE GLYCOL','1,2-ETHANEDIOL',\n",
    "        '(4S)-2-METHYL-2,4-PENTANEDIOL' ,'2-AMINO-2-HYDROXYMETHYL-PROPANE-1,3-DIOL',\n",
    "        '4-(2-HYDROXYETHYL)-1-PIPERAZINE ETHANESULFONIC ACID',\n",
    "        '2-(N-MORPHOLINO)-ETHANESULFONIC ACID',\"ADENOSINE-5'-TRIPHOSPHATE\",\n",
    "        \"NADP NICOTINAMIDE-ADENINE-DINUCLEOTIDE PHOSPHATE\"]\n",
    "    \n",
    "    not_allowed_metals = ['CALCIUM ION','COPPER (II) ION',  'FE (II) ION',\n",
    "                          'FE (III) ION', \"IODIDE ION\",'MAGNESIUM ION',\n",
    "                          'MANGANESE (II) ION','POTASSIUM ION', 'ZINC ION',]\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    # define if ligand is unsuitable/not allowed\n",
    "    unsuitable = False\n",
    "\n",
    "    # get all ligands/non_proteins in pdb\n",
    "    non_proteins = [entity for entity in  mmtf_record.entity_list if entity['type'] == 'non-polymer']\n",
    "\n",
    "    # if len(non_proteins) > 0: # if = 0 => no other molecyles than protein\n",
    "    for entity in non_proteins:\n",
    "        ligand = entity['description']\n",
    "        if ligand  in not_allowed_metals or ligand not in standard_crystallization_ligands:\n",
    "            unsuitable = True\n",
    "\n",
    "    return unsuitable  \n",
    "\n",
    "\n",
    "\n",
    "def rmsd_of_NMR_models_to_high(pdb_name, method):\n",
    "    '''takes all models of NMR model and calculated the RMSD between them \n",
    "    '''\n",
    "    \n",
    "    if 'SOLUTION NMR' in method or 'SOLID-STATE NMR' in method:\n",
    "        try:\n",
    "            traj = md.load(f'/home/trz846/protein_dynamics/data/pdbs_raw/{pdb_name}.pdb')\n",
    "            distances = np.empty((traj.n_frames, traj.n_frames))\n",
    "            for i in range(traj.n_frames):\n",
    "                distances[i] = md.rmsd(traj, traj, i)\n",
    "            max_rmsd = np.max(distances)\n",
    "            if max_rmsd > 0.45:\n",
    "                return True\n",
    "            else: \n",
    "                return False \n",
    "        except:\n",
    "            return True    \n",
    "    else:\n",
    "        return False\n",
    "\n",
    "    \n",
    "    \n",
    "def rmsd_of_NMR_models(pdb_name, method):\n",
    "    '''takes all models of NMR model and calculated the RMSD between them \n",
    "    '''\n",
    "    \n",
    "    if 'SOLUTION NMR' in method or 'SOLID-STATE NMR' in method:\n",
    "        try:\n",
    "            traj = md.load(f'/home/trz846/protein_dynamics/data/pdbs_raw/{pdb_name}.pdb')\n",
    "            distances = np.empty((traj.n_frames, traj.n_frames))\n",
    "            for i in range(traj.n_frames):\n",
    "                distances[i] = md.rmsd(traj, traj, i)\n",
    "            max_rmsd = np.max(distances)\n",
    "            return float(\"{:.2f}\".format(max_rmsd))\n",
    "        except:\n",
    "            return np.nan\n",
    "    else:\n",
    "        return np.nan\n",
    "\n",
    "def get_resolution(mmtf_record, method  ):\n",
    "    '''check if resolution lower than 2.0 AA if experimental method is x-ray\n",
    "    crystallization\n",
    "    '''\n",
    "    \n",
    "    if 'X-RAY DIFFRACTION' in method or 'ELECTRON MICROSCOPY' in method or 'ELECTRON CRYSTALLOGRAPHY' in method : \n",
    "        # can only get resolution for some exp methods \n",
    "        resolution = mmtf_record.resolution \n",
    "        return float(\"{:.2f}\".format(resolution))\n",
    "    else: \n",
    "        return np.nan\n",
    "               \n",
    "def add_prefiltered(pdb_name, all_data):\n",
    "    '''used to create ALL_CATH json database '''\n",
    "    all_data[pdb_name]['prefilter status'] = 'removed'\n",
    "    all_data[pdb_name]['simulation status'] = 'N.A.'\n",
    "    all_data[pdb_name]['simulation reason'] = 'N.A.'\n",
    "    \n",
    "    return all_data\n",
    "\n",
    "\n",
    "def mmtf_fetch(pdb, cache_dir='/home/trz846/graph_transformers/data/ingraham_data/cath/mmtf/'):\n",
    "    \"\"\" Retrieve mmtf record from PDB with local caching. \n",
    "    From J. Ingraham's code \"\"\"\n",
    "    \n",
    "    mmtf_file = cache_dir + pdb + '.mmtf.gz'\n",
    "    if not os.path.isfile(mmtf_file):\n",
    "        url = 'http://mmtf.rcsb.org/v1.0/full/' + pdb + '.mmtf.gz'\n",
    "        mmtf_file = download_cached(url, mmtf_file)\n",
    "\n",
    "    if mmtf_file is not None:\n",
    "        mmtf_record = mmtf.parse_gzip(mmtf_file)\n",
    "    else:\n",
    "        mmtf_record = None  # if pdb obsolete\n",
    "\n",
    "    return mmtf_record\n",
    "\n",
    "def download_cached(url, target_location):\n",
    "    \"\"\" Download with caching. Modified From J. Ingraham's code \"\"\"\n",
    "#     target_dir = os.path.dirname(target_location)\n",
    "#     if not os.path.isfile(target_location):\n",
    "#         if not os.path.exists(target_dir):\n",
    "#             os.makedirs(target_dir)\n",
    "\n",
    "    # Use MMTF for speed\n",
    "    try:\n",
    "        response = urllib.request.urlopen(url)\n",
    "        size = int(float(response.headers['Content-Length']) / 1e3)\n",
    "        print('Downloading {}, {} KB'.format(target_location, size))\n",
    "        with open(target_location, 'wb') as f:\n",
    "            f.write(response.read())\n",
    "    except:\n",
    "        print('pdb is likely obsolete, cannot download:', url)\n",
    "        target_location = None\n",
    "\n",
    "    return target_location\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create dictionary of filtered pdbs from Ingraham's datasplilt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 5iz7\n",
      "\n",
      " ERROR for 5gka\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3jb9\n",
      "\n",
      " ERROR for 3jb9\n",
      "\n",
      " ERROR for 3jb9\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/5d95.mmtf.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain h is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain i is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain n is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain o is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain q is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain s is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain t is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain u is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain y is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain z is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3jcu\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/2kea.mmtf.gz\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/5a0j.mmtf.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 229, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 288, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3j26\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 31, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 38, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 41, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 57, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 58, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 59, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 69, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 97, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 112, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 131, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3jam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 123, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 2n1f\n",
      "\n",
      " ERROR for 3jb8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 239, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain x is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:92: PDBConstructionWarning: WARNING: Chain w is discontinuous at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 5tj5\n",
      "\n",
      " ERROR for 5gae\n",
      "\n",
      " ERROR for 5gae\n",
      "\n",
      " ERROR for 5gae\n",
      "\n",
      " ERROR for 5gae\n",
      "\n",
      " ERROR for 5jzg\n",
      "\n",
      " ERROR for 3jci\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/3fsp.mmtf.gz\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/4otw.mmtf.gz\n",
      "\n",
      " ERROR for 3jc2\n",
      "\n",
      " ERROR for 3zeu\n",
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/4eln.mmtf.gz\n",
      "\n",
      " ERROR for 3jcu\n",
      "\n",
      " ERROR for 3jck\n",
      "\n",
      " ERROR for 5h1s\n",
      "\n",
      " ERROR for 5h1s\n",
      "\n",
      " ERROR for 5ll6\n",
      "\n",
      " ERROR for 3j27\n",
      "\n",
      " ERROR for 3j27\n",
      "\n",
      " ERROR for 5k0u\n",
      "\n",
      " ERROR for 3jam\n",
      "\n",
      " ERROR for 3jam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 166, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pdb is likely obsolete, cannot download: http://mmtf.rcsb.org/v1.0/full/3kwu.mmtf.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:249: PDBConstructionWarning: WARNING: disordered atom found with blank altloc before line 0.\n",
      "\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3j7a\n",
      "\n",
      " ERROR for 3j7a\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 477, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 133, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3j4u\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 2k9y\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 1, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 4, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 5, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 7, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 11, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 15, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 17, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 18, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 24, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n",
      "/home/trz846/anaconda3/envs/graph_transformers/lib/python3.7/site-packages/Bio/PDB/StructureBuilder.py:130: PDBConstructionWarning: WARNING: Residue (' ', 34, ' ') redefined at line 0.\n",
      "  PDBConstructionWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7y\n",
      "\n",
      " ERROR for 3j7a\n",
      "\n",
      " ERROR for 3jcu\n"
     ]
    }
   ],
   "source": [
    "# Define filtering settings \n",
    "MAX_LENGTH = 500\n",
    "MIN_LENGTH = 50 \n",
    "\n",
    "\n",
    "\n",
    "# initiate data container\n",
    "all_data = {}\n",
    "\n",
    "\n",
    "# load list of membrane proteins\n",
    "membrane_proteins = load_membrane_protein_list()\n",
    "\n",
    "\n",
    "# get split into data class from cath types\n",
    "cath_split = './Ingrahams_chain_set_splits.json'\n",
    "with open(cath_split) as f:\n",
    "    pdbs = json.load(f)\n",
    "    \n",
    "    \n",
    "for ds_type in ['train', 'validation', 'test']:\n",
    "    # get all pdbs in each dataset \n",
    "    split_pdbs_w_chains = pdbs[ds_type]\n",
    "    \n",
    "    # remove the chain specific split of data from Ingraham\n",
    "    split_pdbs = [[pdb_chain[:-2],pdb_chain[-1]] for pdb_chain in split_pdbs_w_chains]\n",
    "    \n",
    "    \n",
    "    # loop over all pdbs in given data split\n",
    "    for idx, (pdb_name, chain) in enumerate(split_pdbs):\n",
    "        \n",
    "        all_data[pdb_name] = {'split':ds_type, 'prefilter reasons':[]}\n",
    "        all_data[pdb_name]['cath_nodes'] = pdbs['cath_nodes'][pdb_name+'.'+chain]\n",
    "        \n",
    "        # assume not to filter out, will be changed if filtering conditions are met\n",
    "        filter_out = False\n",
    "\n",
    "        # load structure via mmtf parser \n",
    "        cache_dir = '/home/trz846/graph_transformers/data/ingraham_data/cath/mmtf/'\n",
    "        mmtf_pdb = mmtf_fetch(pdb_name, cache_dir)\n",
    "        try:\n",
    "                    # skip if pdb has become obsolete in pdb database\n",
    "            if  mmtf_pdb is None:\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['size'] = np.nan\n",
    "                all_data[pdb_name]['chains'] = np.nan\n",
    "                all_data[pdb_name]['method'] = 'N.A'\n",
    "                all_data[pdb_name]['prefilter reasons'].append('obsolete')\n",
    "                filter_out = True\n",
    "                continue\n",
    "\n",
    "            # load as biopython object as works better for polychains and missing aa's\n",
    "            biopython_structure =  get_biopython_structure(mmtf_pdb)\n",
    "\n",
    "            # add basics about pdb\n",
    "            all_data[pdb_name]['size'] = get_size(biopython_structure)\n",
    "            all_data[pdb_name]['chains'] = polymeric(biopython_structure)\n",
    "            methods = mmtf_pdb.experimental_methods\n",
    "            all_data[pdb_name]['method'] = methods\n",
    "    #         all_data[pdb_name]['rmsd NMR'] = rmsd_of_NMR_models(pdb_name, methods)\n",
    "    #         all_data[pdb_name]['resolution'] =  get_resolution(mmtf_pdb, methods )\n",
    "\n",
    "            if membrane_protein(pdb_name):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('membrane protein')\n",
    "                filter_out = True\n",
    "            if unsuitable_experimental_method(mmtf_pdb):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('unsuitable exp method')\n",
    "                filter_out = True\n",
    "            if low_resolution(mmtf_pdb): \n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('low resolution')\n",
    "                filter_out = True\n",
    "\n",
    "            if rmsd_of_NMR_models_to_high(pdb_name, all_data[pdb_name]['method'] ):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('NMR rmsd to high')\n",
    "                filter_out = True\n",
    "\n",
    "            if unknown_aa(mmtf_pdb):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('unknown_aa')\n",
    "                filter_out = True\n",
    "\n",
    "            if polymeric_with_single_chain_too_short(biopython_structure):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('one chain too short')\n",
    "                filter_out = True\n",
    "            if missing_aa_inside_protein(biopython_structure):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('gap in chain')\n",
    "                filter_out = True\n",
    "            if unsuitable_ligands(mmtf_pdb):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('unsuitable ligand')\n",
    "                filter_out = True\n",
    "            if too_big(biopython_structure, max_len = MAX_LENGTH):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('too big')\n",
    "                filter_out = True\n",
    "            if too_small(biopython_structure, min_len = MIN_LENGTH):\n",
    "                all_data = add_prefiltered(pdb_name, all_data)\n",
    "                all_data[pdb_name]['prefilter reasons'].append('too small')\n",
    "                filter_out = True\n",
    "\n",
    "            all_data[pdb_name]['rmsd NMR'] = rmsd_of_NMR_models(pdb_name, methods)\n",
    "            all_data[pdb_name]['resolution'] =  get_resolution(mmtf_pdb, methods )\n",
    "\n",
    "            if not filter_out:\n",
    "                all_data[pdb_name]['prefilter status'] = 'kept'\n",
    "                all_data[pdb_name]['simulation status'] = '-'\n",
    "                all_data[pdb_name]['unsuccessfull reason'] = '-'\n",
    "                all_data[pdb_name]['size'] = get_size(biopython_structure)\n",
    "                all_data[pdb_name]['chains'] = polymeric(biopython_structure)\n",
    "                all_data[pdb_name]['method'] = mmtf_pdb.experimental_methods\n",
    "        except:\n",
    "            all_data = add_prefiltered(pdb_name, all_data)\n",
    "            all_data[pdb_name]['prefilter reasons'] = 'error in reading pdb'\n",
    "            all_data[pdb_name]['size'] = np.nan\n",
    "            all_data[pdb_name]['chains'] = np.nan\n",
    "            all_data[pdb_name]['method'] = 'N.A'\n",
    "            print(f'\\n ERROR for {pdb_name}')\n",
    "       \n",
    "    \n",
    "    \n",
    "    \n",
    "with open('CATH4.2_INITIAL_STATUS.json', 'w') as fp:\n",
    "    json.dump(all_data, fp)  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "graph_transformers",
   "language": "python",
   "name": "graph_transformers"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
